src/coffea/dataset_tools/filespec.py

```diff
@@ -0,0 +1,441 @@
+    format: Literal["root"] = "root"
+
+
+class CoffeaROOTFileSpecOptional(ROOTFileSpec):
+    num_entries: Annotated[int, Field(ge=0)] | None = None
+    uuid: str | None = None
+
+
+class CoffeaROOTFileSpec(CoffeaROOTFileSpecOptional):
+    steps: Annotated[list[StepPair], Field(min_length=1)] | StepPair
+    num_entries: Annotated[int, Field(ge=0)]
+    uuid: str
+
+
+class ParquetFileSpec(GenericFileSpec):
+    object_path: None = None
+    format: Literal["parquet"] = "parquet"
+
+
+class CoffeaParquetFileSpecOptional(ParquetFileSpec):
+    num_entries: Annotated[int, Field(ge=0)] | None = None
+    uuid: str | None = None
+
+
+class CoffeaParquetFileSpec(CoffeaParquetFileSpecOptional):
+    steps: Annotated[list[StepPair], Field(min_length=1)] | StepPair
+    num_entries: Annotated[int, Field(ge=0)]
+    uuid: str
+    # is_directory: Literal[True, False] #identify whether it's a directory of parquet files or a single parquet file, may be useful or necessary to distinguish
+
+
+class DictMethodsMixin:
+    def __getitem__(self, key: str) -> str:
+        return self.root[key]
+
+    def __setitem__(self, key: str, value: str):
+        self.root[key] = value
+
+    def __delitem__(self, key: str):
+        del self.root[key]
+
+    def __len__(self) -> int:
+        return len(self.root)
+
+    def __iter__(self) -> Iterable[str]:
+        return iter(self.root)
+
+    def keys(self) -> Iterable[str]:
+        return self.root.keys()
+
+    def values(self) -> Iterable[str]:
+        return self.root.values()
+
+    def items(self) -> Iterable[tuple[str, str]]:
+        return self.root.items()
+
+    def get(self, key: str, default=None) -> str | None:
+        return self.root.get(key, default)
+
+    def pop(self, key: str, default=...):
+        return self.root.pop(key, default)
+
+    def update(self, other=None, **kwargs):
+        self.root.update(other, **kwargs)
+
+
+class InputFilesMixin:
+    @computed_field
+    @property
+    def format(self) -> str:
+        """Identify the format of the files in the dictionary."""
+        union = set()
+        identified_formats_by_name = {
+            k: identify_file_format(k) for k, v in self.root.items()
+        }
+        stored_formats_by_name = {
+            k: v.format for k, v in self.root.items() if hasattr(v, "format")
+        }
+        if not all(
+            k in identified_formats_by_name and identified_formats_by_name[k] == v
+            for k, v in stored_formats_by_name.items()
+        ):
+            raise ValueError(
+                f"identified formats and stored formats do not match: identified formats: {identified_formats_by_name}, stored formats: {stored_formats_by_name}"
+            )
+        union.update(identified_formats_by_name.values())
+        if len(union) == 1:
+            return union.pop()
+        return "|".join(union)
+
+    @model_validator(mode="before")
+    def preproc_data(cls, data: Any) -> Any:
+        data = copy.deepcopy(data)
```

Comment:
Is this to convert data that is not in the canonical form? Would it be possible to detect that and then decide to copy? Would it make more sense as a classmethod `from_legacy_dict()`?